api/src/main/java/org/infinispan/api/async/Cache.java

```diff
@@ -0,0 +1,197 @@
+
+   /**
+    * The configuration for this cache.
+    *
+    * @return
+    */
+   CompletionStage<CacheConfiguration> configuration();
+
+   /**
+    * Returns a new instance of this cache with the supplied {@link Flag}s
+    * @param flags
+    * @return
+    */
+   Cache<K, V> withFlags(EnumSet<Flag> flags);
+
+   /**
+    * Returns a new instance of this cache with the supplied {@link Flag}s
+    * @param flags
+    * @return
+    */
+   Cache<K, V> withFlags(Flag... flags);
+
+   /**
+    * Get the value of the Key if such exists
+    *
+    * @param key
+    * @return the value
+    */
+   CompletionStage<V> get(K key);
+
+   /**
+    * Insert the key/value if such key does not exist
+    *
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<Boolean> putIfAbsent(K key, V value);
+
+   /**
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<V> put(K key, V value);
+
+   /**
+    * @param key
+    * @param value
+    * @return
+    */
+   CompletionStage<Void> set(K key, V value);
+
+   /**
+    * Delete the key
+    *
+    * @param key
+    * @return whether the entry was removed.
+    */
+   CompletionStage<Boolean> remove(K key);
+
+   /**
+    * Delete the key, returning its value
+    *
+    * @param key
+    * @return whether the entry was removed.
+    */
+   CompletionStage<V> getAndRemove(K key);
+
+   /**
+    * Retrieve all keys
+    *
+    * @return
+    */
+   Flow.Publisher<K> keys();
+
+   /**
+    * Retrieve all entries
+    *
+    * @return
+    */
+   Flow.Publisher<? extends Map.Entry<K, V>> entries();
+
+   /**
+    * @param entries
+    * @return Void
+    */
+   CompletionStage<Void> putAll(Map<K, V> entries);
+
+   /**
+    * Retrieves the entries for the specified keys.
+    *
+    * @param keys
+    * @return
+    */
+   Flow.Publisher<? extends Map.Entry<K, V>> getAll(Set<K> keys);
+
+   /**
+    * Removes a set of keys. Returns the keys that were removed.
+    *
+    * @param keys
+    * @return
+    */
+   Flow.Publisher<K> removeAll(Set<K> keys);
+
+   /**
+    * Removes a set of keys. Returns the keys that were removed.
+    *
+    * @param keys
+    * @return
+    */
+   Flow.Publisher<? extends Map.Entry<K, V>> getAndRemoveAll(Set<K> keys);
+
+   /**
+    * Estimate the size of the store
+    *
+    * @return Long, estimated size
+    */
+   CompletionStage<Long> estimateSize();
+
+   /**
+    * Clear the store. If a concurrent operation puts data in the store the clear might not properly work
+    *
+    * @return Void
+    */
+   CompletionStage<Void> clear();
+
+   /**
+    * Executes the query and returns an iterable with the entries
+    *
+    * @param query query String
+    * @return Publisher reactive streams
+    */
+   <R> CompletionStage<QueryResult<R>> find(String query);
+
+   /**
+    * @param listener
+    * @return
+    */
+   CompletionStage<AsyncListenerHandle<KeyValueListener>> listen(KeyValueListener listener);
```

Comment:
We would have to make sure removal of a listener works when invoked concurrently with it being registered then. I don't think that is supported atm. But it would make the API look nicer so I can get behind that.